### PR TITLE
FIX: Do not evaluate callable defaults during tab-completion

### PR DESCRIPTION
**Pull request by danielhollas**
In #6140 we've tried to speed up verdi invocation by lazy loading config / profile. Unfortunately, the configuration is still being loaded during tab-completion.

After fair amount of going through the code in both click and aiida, I now think this is a bug in click, see https://github.com/pallets/click/issues/2614. I've devised a fix that passes the click test suite so it seems that the current behavior is unintented.

I will submit a PR to click, but given the speedup that we stand to gain (~50ms) for such a time-sensitive thing as tab-completion, I think it is worth fixing here for the next aiida-core release, which I suspect will happen before the next click release.

I've verified that with this fix, the profile is indeed not being loaded during tab-completion, by stucking `raise ValueError` in the `aiida.manage.configuration.get_config()` and observing that it raises on main and does not raise here during tab completion.

It would be great to have a regression test for this, but I am not sure how to do it. Here's how click tests it: https://github.com/pallets/click/blob/main/tests/test_shell_completion.py

### CODE REVIEW

File: aiida/cmdline/params/options/interactive.py [hunk 1]
```diff
@@ -158,7 +158,7 @@ def get_default(self, ctx: click.Context, call: bool = True) -> t.Optional[t.Uni
         if self._contextual_default is not None:
             default = self._contextual_default(ctx)
         else:
-            default = super().get_default(ctx)
+            default = super().get_default(ctx, call=call)
```
Comment (danielhollas) [inline aiida/cmdline/params/options/interactive.py:161]: Unrelated fix

**Comment by danielhollas** [comment]
@sphuber this is ready for review

File: tests/cmdline/commands/test_setup.py [hunk 2]
```diff
@@ -206,7 +206,7 @@ def test_setup_profile_uuid(self):
 
         profile_name = 'profile-copy'
         user_email = 'some@email.com'
-        profile_uuid = str(uuid.uuid4)
+        profile_uuid = str(uuid.uuid4())
```
[the remark this answers is not in the document]
Comment (danielhollas) [reply tests/cmdline/commands/test_setup.py:209]: Will remove this since it will be fixed in the other PR.

File: aiida/cmdline/params/options/callable.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    """
+
+    def get_default(self, ctx: click.Context, call: bool = True) -> t.Optional[t.Union[t.Any, t.Callable[[], t.Any]]]:
+        """provides the functionality of :meth:`click.Option.get_default`,
+        but ensures we do not evaluate callable defaults when in tab-completion context."""
+        if ctx.resilient_parsing:
+            # This is done in interactive.py:InteractiveOption as well
+            return None
+            # Another possibility
+            # return super().get_default(ctx=ctx, call=False)
```
Comment (sphuber) [inline aiida/cmdline/params/options/callable.py:36]:
```suggestion
        if ctx.resilient_parsing:
            return None
```
I would just keep the version that we also keep in the InteractiveOption. Looked at the source and the `Option` class has some additional logic for non-boolean flags that does not seem to take `ctx.resilient_parsing` into account. Unlikely we would hit this, but I think returning `None` does the job and prevents any other complexity

File: tests/cmdline/commands/test_setup.py [hunk 4]
```diff
@@ -206,6 +206,7 @@ def test_setup_profile_uuid(self):
 
         profile_name = 'profile-copy'
         user_email = 'some@email.com'
+
```
Comment (sphuber) [inline tests/cmdline/commands/test_setup.py:209]:
```suggestion
```

File: aiida/cmdline/params/options/main.py [hunk 5]
```diff
@@ -145,6 +146,7 @@ def set_log_level(_ctx, _param, value):
     'profile',
     type=types.ProfileParamType(),
     default=defaults.get_default_profile,
+    cls=CallableDefaultOption,
```
Comment (sphuber) [inline aiida/cmdline/params/options/main.py:149]: Is this really the only option that has an expensive callable default?

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @danielhollas . As for the test, maybe you could try something like the following:

```python
# -*- coding: utf-8 -*-
###########################################################################
# Copyright (c), The AiiDA team. All rights reserved.                     #
# This file is part of the AiiDA code.                                    #
#                                                                         #
# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
# For further information on the license, see the LICENSE.txt file        #
# For further information please visit http://www.aiida.net               #
###########################################################################
# pylint: disable=redefined-outer-name
"""Tests for the :mod:`aiida.cmdline.params.options.callable` module."""
import pytest

from click.shell_completion import ShellComplete

from aiida.cmdline.commands.cmd_verdi import verdi


def _get_completions(cli, args, incomplete):
    comp = ShellComplete(cli, {}, cli.name, '_CLICK_COMPLETE')
    return comp.get_completions(args, incomplete)


@pytest.fixture
def unload_config():
    """Temporarily unload the config by setting ``aiida.manage.configuration.CONFIG`` to ``None``."""
    from aiida.manage import configuration

    config = configuration.CONFIG

    try:
        configuration.CONFIG = None
        yield
    finally:
        configuration.CONFIG = config


@pytest.mark.usefixtures('unload_config')
def test_callable_default_resilient_parsing():
    """Test that tab-completion of ``verdi`` does not evaluate defaults that load the config, which is expensive."""
    from aiida.manage import configuration

    assert configuration.CONFIG is None
    [c.value for c in _get_completions(verdi, [], '')]
    assert configuration.CONFIG is None
```

This fails for the `main` branch as it should. If it passes on your branch, I would say this provides some assurance it is working as intended

File: aiida/cmdline/params/options/main.py (hunk 5 above)
Comment (danielhollas) [reply aiida/cmdline/params/options/main.py:149]: It's the only place where the config/profile is loaded, all the other use the InteractiveOption where this is already handled. But you are right that there likely are other expensive defaults, but I plan to look into this in a followup PR where I will also look at the timings more closely.

**Comment by sphuber** [comment]
> [One more place that loaded the config](https://github.com/aiidateam/aiida-core/pull/6144/commits/cade228136fcd4e11cc8ab6d55411322eebc5faa)

Seems like the test is doing its job :+1:

File: tests/cmdline/params/options/test_callable.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    return comp.get_completions(args, incomplete)
+
+
+@pytest.fixture
+def unload_config():
+    """Temporarily unload the config by setting ``aiida.manage.configuration.CONFIG`` to ``None``."""
+    from aiida.manage import configuration
+
+    config = configuration.CONFIG
+    configuration.CONFIG = None
```
Comment (danielhollas) [inline tests/cmdline/params/options/test_callable.py:29]: NOTE: I have removed the try-except block, I don't think it is necessary, `pytest` should ensure that the fixture is run to completion after the test, unless the fixture itself excepts before the yield point, but here we only have two assignments.

https://docs.pytest.org/en/latest/how-to/fixtures.html#teardown-cleanup-aka-fixture-finalization

https://docs.pytest.org/en/latest/how-to/fixtures.html#safe-teardowns

File: tests/cmdline/params/options/test_callable.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+
+@pytest.mark.usefixtures('unload_config')
+def test_callable_default_resilient_parsing():
+    """Test that tab-completion of ``verdi`` does not evaluate defaults that load the config, which is expensive."""
+    from aiida.manage import configuration
+
+    assert configuration.CONFIG is None
+    completions = [c.value for c in _get_completions(verdi, [], '')]
+    assert 'help' in completions
```
Comment (danielhollas) [inline tests/cmdline/params/options/test_callable.py:41]: I noticed that we do not test autocompletion anywhere else in the test suite. I'll try adding more tests in a separate PR, for now I added at least this simple assert.
(also to shutup pylint which was complaining about unassigned expression)

**Review by danielhollas — COMMENTED** [review]
Haha, indeed, thank you very much for the test! This is now ready from my side.

File: aiida/cmdline/params/options/main.py [hunk 8]
```diff
@@ -110,8 +111,10 @@ def set_log_level(_ctx, _param, value):
     log.CLI_ACTIVE = True
 
     # If the value is ``None``, it means the option was not specified, but we still configure logging for the CLI
+    # However, we skip this when we are in a tab-completion context.
     if value is None:
-        configure_logging()
+        if not _ctx.resilient_parsing:
+            configure_logging()
```
Comment (sphuber) [inline aiida/cmdline/params/options/main.py:117]: I am a bit surprised by this change. This function `set_log_level` is only assigned as the callback of the `VERBOSITY` option. I don't think this is supposed to be called during tab-completion anyway. I just tested this and it indeed doesn't seem to be called during tab-completion. Was this the part of the code that caused the new test to fail? Do you understand why?

Comment (danielhollas) [reply aiida/cmdline/params/options/main.py:117]: Good catch, sorry for not being clear, I am also confused, but you can try when you remove it the test fails. But when I test the completion on the actual command line it is not called. Maybe the click function used in the test is not exactly the one that gets called?? Btw: I was testing on BASH, wonder if other shells may behave differently.

Comment (danielhollas) [reply aiida/cmdline/params/options/main.py:117]: Perhaps there is some weird interaction with the test suite. Not sure if it is worth deeper investigation, since the change itself seems like an okay thing to do on its own.

Comment (sphuber) [reply aiida/cmdline/params/options/main.py:117]: Figured it out. I tested whether this was being called when actually tab-completing `verdi` by adding a `print` statement. Since that print statement never showed up, I concluded that the function wasn't being called. But that is not true. It _was_ actually called, but during tab-completion, all output to `sys.stdout` is captured and so I didn't see anything. Printing to `sys.stderr` would actually show, or simply raising an exception would confirm the function was being called.

**Review by sphuber — APPROVED** [review]
Thanks @danielhollas
